src/diffusion/train_utils.py

```diff
@@ -0,0 +1,96 @@
+import torch
+import torch.nn.functional as F
+from torch import nn
+from tqdm import tqdm
+
+from schedules import cosine_beta_schedule
+from schedules import linear_beta_schedule
+
+
+def configure_schedule(steps: int, schedule: str):
+    if schedule == "linear":
+        betas = linear_beta_schedule(steps)
+    elif schedule == "cosine":
+        betas = cosine_beta_schedule(steps)
+    else:
+        raise NotImplementedError(f"TODO: implement {schedule} schedule")
+
+    alphas = 1 - betas
+    alphas_cumprod = torch.cumprod(alphas, dim=0)
+    alphas_cumprod_prev = F.pad(alphas_cumprod[:-1], (1, 0), value=1.0)
+    return alphas_cumprod_prev
+
+
+def get_diffusion_variables(
+    model,
+    gt,
+    max_gen: int,
+    diffusion_steps: int,
+    alphas_cumprod_prev: torch.Tensor,
+    device: str,
+):
+    pad_emb = model.get_embeddings(torch.LongTensor([0]))
```

Comment:
- Pad may have another index. I think you specify it by checking the id in the tokenizer.
- I suggest moving this function inside `DiffusionTransformer` (btw, `Seq2SeqDiffusionTransformer` would suit better). You can save the pad vector to the module state (self._pad_emb = ...) and reuse it.